### PR TITLE
Refactoring of the introspection functionalities

### PR DESCRIPTION
**Pull request by novaknole**

### CODE REVIEW

**Comment by novaknole** [comment]
The PR is ready to be merged. @Rekard0 could you look into the failing tests on the subgraph side ?

File: packages/contracts/contracts/core/DAO.sol [hunk 1]
```diff
@@ -233,4 +237,13 @@ contract DAO is IDAO, Initializable, UUPSUpgradeable, PermissionManager, ERC1271
...
     }
+
+    /// @inheritdoc IDAO
+    function registerStandardAndCallback(
+        bytes4 _interfaceId,
+        bytes4 _callbackSig,
+        bytes4 _magicNumber
+    ) external override auth(address(this), REGISTER_STANDARD_PERMISSION_ID) {
+        _registerStandardAndCallback(_interfaceId, _callbackSig, _magicNumber);
+    }
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/DAO.sol:248]: I still don't understand in which situation we need AdaptiveERC165 and its callback functionality.

Why does our DAO want to add more interfaces? 

If there is a reason, I would just use OZs `ERC165Storage` (see https://docs.openzeppelin.com/contracts/4.x/api/utils#ERC165Storage and https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/introspection/ERC165Storage.sol) that can register multiple interfaces and remove `AdaptiveERC165`.
We would need to add a `register` function here in this case.

If the DAO doesn't need this, `DAO` should just inherit from OZs `ERC165` and have only one interface.

File: packages/contracts/contracts/core/permission/PermissionManager.sol [hunk 2]
```diff
@@ -343,7 +343,10 @@ contract PermissionManager is Initializable {
     /// @notice A modifier to be used to check permissions on a target contract.
     /// @param _where The address of the target contract for which the permission is required.
     /// @param _permissionId The permission identifier required to call the method this modifier is applied to.
-    function _auth(address _where, bytes32 _permissionId) private {
+    function _auth(address _where, bytes32 _permissionId) private view {
+        // TODO: will we need to emit `who` as well ? below `who: msg.sender` 
+        // is not actul who that we grant the permission in the end, it's the sender
+        // that initiations permission.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/permission/PermissionManager.sol:349]: Let's not modify the permission manager in this PR and do it in a separate PR, e.g., https://github.com/aragon/core/pull/90.
Please put the comments there.

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 3]
```diff
@@ -254,4 +254,11 @@ contract PluginRepo is
         override
         auth(address(this), UPGRADE_PERMISSION_ID)
     {}
+
+    /// @notice adds a IERC165 to check whether contract supports GovernanceERC20 interface or not.
+    /// @dev See {ERC165Upgradeable-supportsInterface}.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/plugin/PluginRepo.sol:259]: This description is incorrect + this kind of referencing does not work

File: packages/contracts/contracts/registry/InterfaceBasedRegistry.sol [hunk 4]
```diff
@@ -64,6 +64,7 @@ abstract contract InterfaceBasedRegistry is DaoAuthorizableUpgradeable, UUPSUpgr
 
         if (entries[_registrant]) revert ContractAlreadyRegistered({registrant: _registrant});
 
+        // TODO: if we keep using this contract, instead of AdaptiveERC165, use ERC165Checker from OZ.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/registry/InterfaceBasedRegistry.sol:67]: Let's do it now. Otherwise, we would not be able to delete AdaptiveERC165 if we realize that we don't need it.

**Review by Michael-A-Heuer — DISMISSED** [review]
I don't get what we need `AdaptiveERC165` and its callback functionality for. 
If nothing is speaking against it I would replace it by [OZs `ERC165Storage`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/introspection/ERC165Storage.sol) or, more simply, [OZs `ERC165`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/introspection/ERC165.sol) instead and save gas.

File: packages/contracts/test/plugin/plugin-repo.ts [hunk 5]
```diff
@@ -90,7 +90,7 @@ describe('PluginRepo', function () {
 
     // TODO: GIORGI fix after the repo is fixed...
     await expect(
-      pluginRepo.createVersion([1, 0, 0], adaptiveERC165.address, emptyBytes)
+      pluginRepo.createVersion([1, 0, 0], pluginRepo.address, emptyBytes)
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/test/plugin/plugin-repo.ts]: Got some time to workaround this, I had to safely add piece of code in PluginRepo which was the reason this test was skipped. Not anymore though.

File: packages/contracts/contracts/core/permission/PermissionManager.sol (hunk 2 above)
Comment (novaknole) [reply packages/contracts/contracts/core/permission/PermissionManager.sol]: Noted. Though, I will leave `'view` keyword.  Not a big overhead, comment removed + test that was brought for our new bug to fix it removed as well (it's in another PR).

File: packages/contracts/contracts/registry/InterfaceBasedRegistry.sol (hunk 4 above)
Comment (novaknole) [reply packages/contracts/contracts/registry/InterfaceBasedRegistry.sol]: done

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 6]
```diff
@@ -95,24 +98,20 @@ contract PluginRepo is
         address _pluginSetup,
         bytes calldata _contentURI
     ) external auth(address(this), CREATE_VERSION_PERMISSION_ID) {
-        // Check if `_pluginSetup` is a `PluginSetup` contract
-        if (!Address.isContract(_pluginSetup)) {
-            revert InvalidContractAddress({invalidContract: _pluginSetup});
+        // In a case where _pluginSetup doesn't contain supportsInterface,
+        // but contains fallback, that doesn't return anything(most cases)
+        // the below approach aims to still return custom error which not possible with try/catch..
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:103]: Don't we want to keep checking that Plugin Setup is a contract?

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 7]
```diff
@@ -95,24 +98,20 @@ contract PluginRepo is
...
-        //     PluginSetup(_pluginSetup).supportsInterface(
-        //         PluginSetup.PLUGIN_FACTORY_INTERFACE_ID
-        //     )
-        // returns (bool result) {
-        //     if (!result) {
-        //         revert InvalidPluginSetupInterface({invalidPluginSetup: _pluginSetup});
-        //     }
-        // } catch {
-        //     revert InvalidPluginSetupContract({invalidPluginSetup: _pluginSetup});
-        // }
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:114]: Why ceasing to check this?

File: packages/contracts/contracts/test/DAOMock.sol [hunk 8]
```diff
@@ -69,4 +69,10 @@ contract DAOMock is IDAO, PermissionManager {
     ) external pure override returns (bytes4) {
         return 0x0;
     }
+
+    function registerStandardCallback(
```
Comment (brickpop) [inline packages/contracts/contracts/test/DAOMock.sol:73]: What's the main goal behind this new function on the DAO contract?

File: packages/contracts/contracts/plugin/PluginRepo.sol (hunk 6 above)
Comment (novaknole) [reply packages/contracts/contracts/plugin/PluginRepo.sol:103]: The `functionCall` checks it implicitly.

File: packages/contracts/contracts/plugin/PluginRepo.sol (hunk 7 above)
Comment (novaknole) [reply packages/contracts/contracts/plugin/PluginRepo.sol:114]: This had one problem which I didn't like (the problem existed more in EVM level limitation from solidity with try/catch !

File: packages/contracts/contracts/test/DAOMock.sol (hunk 8 above)
Comment (novaknole) [reply packages/contracts/contracts/test/DAOMock.sol:73]: This dynamically can register callbacks (such as `onERC721Received` which can be found on ERC721 contract) so dao can receive erc721 e.x

File: packages/contracts/contracts/core/DAO.sol (hunk 1 above)
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/core/DAO.sol]: Resolved, by discussions with @novaknole and merge of https://github.com/aragon/core/pull/112

**Comment by Michael-A-Heuer** [comment]
I didn't check for test coverage and just fixed the tests that were failing already before my review suggestion PR #112.

File: packages/contracts/contracts/core/permission/PermissionManager.sol (hunk 2 above)
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/core/permission/PermissionManager.sol]: Ok

File: packages/contracts/contracts/core/DAO.sol [hunk 9]
```diff
@@ -2,22 +2,31 @@
...
 /// @dev Public API of the Aragon DAO framework.
-contract DAO is IDAO, Initializable, UUPSUpgradeable, PermissionManager, ERC1271, AdaptiveERC165 {
+contract DAO is
+    IDAO,
+    Initializable,
+    UUPSUpgradeable,
+    PermissionManager,
+    ERC1271,
+    ERC165StorageUpgradeable,
+    CallbackHandler
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/DAO.sol:28]: @novaknole Is the ordering ok?

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 10]
```diff
@@ -21,8 +23,10 @@ contract PluginRepo is
     Initializable,
     UUPSUpgradeable,
     PermissionManager,
-    AdaptiveERC165
+    ERC165Upgradeable
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/plugin/PluginRepo.sol:26]: @novaknole Is the ordering ok?

File: packages/contracts/contracts/test/CallbackHandlerHelperMock.sol [hunk 11]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.10;
+
+contract CallbackHandlerMockHelper {
+    address callbackHandlerMockAddr;
+
+    event CallbackReceived(bytes32 b);
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:8]: Remove unused event
```suggestion
```

File: packages/contracts/contracts/tokens/GovernanceWrappedERC20.sol [hunk 12]
```diff
@@ -23,7 +23,7 @@ import "../core/IDAO.sol";
 /// @dev This contract intentionally has no public mint functionality because this is the responsibility of the underlying [ERC-20](https://eips.ethereum.org/EIPS/eip-20) token contract.
 contract GovernanceWrappedERC20 is
     Initializable,
-    AdaptiveERC165,
+    ERC165Upgradeable,
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/tokens/GovernanceWrappedERC20.sol:26]: @novaknole Is the ordering ok?

File: packages/contracts/contracts/registry/InterfaceBasedRegistry.sol [hunk 13]
```diff
@@ -64,7 +65,7 @@ abstract contract InterfaceBasedRegistry is DaoAuthorizableUpgradeable, UUPSUpgr
 
         if (entries[_registrant]) revert ContractAlreadyRegistered({registrant: _registrant});
 
-        try AdaptiveERC165(_registrant).supportsInterface(targetInterfaceId) returns (bool result) {
+        try IERC165(_registrant).supportsInterface(targetInterfaceId) returns (bool result) {
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/registry/InterfaceBasedRegistry.sol:68]: Is it ok to do it like that @novaknole ?

**Review by Michael-A-Heuer — COMMENTED** [review]
Looks good, I can approve if the remaining conversations are resolved.

**Review by Michael-A-Heuer — APPROVED** [review]
LGTM

**Review by Rekard0 — APPROVED** [review]
